File: python/201-250/210. Course Schedule II.py
import collections


class Solution(object):
    def findOrder(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: List[int]
        """
<<<<<<< HEAD
        graph = collections.defaultdict(list)
        firstlevel = [True] * numCourses
        for pair in prerequisites:
            firstlevel[pair[0]] = False
            graph[pair[1]].append(pair[0])
        visited = collections.defaultdict(bool)
        stack = []
        def dfs(cur):
            if len(graph[cur]) == 0:
                stack.append(cur)
                return 
            for node in graph[cur]:
                print node
                if not visited[node]:
                    visited[node] = True
                    dfs(node)
            stack.append(cur)
        for i in range(numCourses):
            if firstlevel[i]:
                visited[i] = True
                dfs(i)
        return stack[::-1]

sol = Solution()
num = 4
pre = [[1,0],[2,1],[3,2],[1,3]]
print sol.findOrder(num, pre)
=======
        dic = collections.defaultdict(list)
        graph = collections.defaultdict(list)
        for i, j in prerequisites:
            dic[i].append(j)
            graph[j].append(i)
        stack = [i for i in xrange(numCourses) if not dic[i]]
        res = []
        while stack:
            node = stack.pop()
            res.append(node)
            for i in graph[node]:
                dic[i].remove(node)
                if not dic[i]:
                    stack.append(i)
            del dic[node]
        return res if not dic else []





sol = Solution()
num = 2
pre = [[1, 0]]
print sol.findOrder(num, pre)
>>>>>>> e1f5edd4b49e5db2d6a60eb8c212da951d007d4b

File: cpp/201-250/210. Course Schedule II.cc
class Solution {
private:
    vector<int> res;
public:
    vector<int> findOrder(int numCourses, vector<pair<int, int>>& prerequisites) {
        if (numCourses == 0)
            return res;
        vector<bool> onpath(numCourses, false);
        vector<unordered_set<int>> graph = make_graph(numCourses, prerequisites);
        bool nocycle;
        for (int i = 0; i < numCourses; i++) {
            vector<bool> visited(numCourses, false);
            nocycle = dfs(i, graph, visited, onpath);
            if (!nocycle){
                res.erase(res.begin(), res.end());
                return res;
            }
        }
        return res;
    }
    vector<unordered_set<int>> make_graph(int numCourses, vector<pair<int, int>>& prerequisites) {
        vector<unordered_set<int>> graph(numCourses);
        for (auto it : prerequisites) {
            graph[it.first].insert(it.second);
        }
        return graph;
    }
    bool dfs(int node, vector<unordered_set<int>> &graph, vector<bool> &visited, vector<bool> &onpath) {
        if (onpath[node])
            return true;
        if (visited[node])
            return false;
        visited[node] = true;
        for (int i : graph[node]) {
            bool res = dfs(i, graph, visited, onpath);
            if (!res)
                return false;
        }
        res.push_back(node);
        onpath[node] = true;
        return true;
    }
};